### ISSUE TITLE
Unable To Install dotnet35sp1

### ISSUE THREAD

**Issue by CodingKoopa** [Winetricks/winetricks#810]
Winetricks sometimes fails to install the `dotnet35sp1` verb.

To get solid test data, I ran the `dotnet35sp1` verb 5 times on clean prefixes with reboots in between:
- Test with Wine 2.21
- Same
- Test with Wine 2.21 Staging
- Same
- Same

The command used to run each is `export WINEARCH=win32 && export WINEPREFIX=~/.wine.${NUMBER} && wineboot &> wineboot.log && winetricks dotnet35sp1 &> winetricks-dotnet35sp1.log`, with NUMBER being the test number. For each of these tests, I got as much data as I could to try and see what the problem is.

# Summary
This section summarizes the overall conclusions I made from the test data. This is put above the tests themselves because they, when together, get to be quite lengthy.
## Wine Stable
On Wine Stable, the verb failed in both of my tests, while installing the .NET 2.0 SP2 runtime, due to the installer not detecting the .NET 2.0 runtime.

## Wine Staging
On Wine Staging, the verb failed in one of my tests for an unknown reason. In the other two tests, a fatal error in either `mscorsvw.exe` or an unidentified program was encountered while installing the .NET 2.0 SP1 and .NET 2.0 SP2 runtimes, always because of an unimplemented function, `mscoree.dll.CorIsLatestSvc`. However, the verb still succeded despite these errors.

# Tests
This section has the test data itself. Each attempt builds off the last unless noted otherwise, with things differing from the last attempt bolded.
## First Attempt
[Wineboot Log](https://github.com/Winetricks/winetricks/files/1502557/wineboot.log)
[Winetricks Log](https://github.com/Winetricks/winetricks/files/1502556/winetricks-dotnet35sp1.log)
### Program Versions
- `wine`: `wine-2.21`
- `winetricks`: `20171018`

### Installations
- .NET 2.0: Success.
- .NET 3.0: Success.
- .NET 2.0 SP1: Success, with errors. Before running the installer, the verb hung while setting the Windows version to `win2k`, at line 1019 of the log. To get around this, I started the task manager with `export WINEARCH=win32 && export WINEPREFIX=~/.wine.1 && wine taskmgr`, and ended every process manually, with the task manager itself last.
![1](https://user-images.githubusercontent.com/13039555/33224274-a7bf6908-d135-11e7-8df3-5762ff32dfef.png)
- XPS Essentials Pack: Success.
- .NET 3.0 SP1: Success.
- .NET 3.5: Success.
- .NET 2.0 SP2: Failure. While running the installer, at line 2005 of the log, the Microsoft .NET Services Installation Utility encountered a fatal error. I clicked the `Don't Send` button. After this, the installation started removing files, and when finished opened up another fatal error message box, at line 2019 of the log. I clicked the `Don't Send Report` button. The main installation window now showed the error text, and link to the error log, saying that the .NET 2.0 runtime was not detected.
![2](https://user-images.githubusercontent.com/13039555/33224275-ad7edd2e-d135-11e7-9fee-71f87469bdaa.png)
![3](https://user-images.githubusercontent.com/13039555/33224276-aebdcf42-d135-11e7-9a03-bbb90be56b2d.png)
![4](https://user-images.githubusercontent.com/13039555/33224277-afbd4472-d135-11e7-9840-2f2ab66ceaa6.png)

## Second Attempt
[Winetricks Log](https://github.com/Winetricks/winetricks/files/1502558/winetricks-dotnet35sp1.log)
### Program Versions
- `wine`: `wine-2.21`
- `winetricks`: `20171018`

### Installations
- .NET 2.0: Success.
- .NET 3.0: Success.
- .NET 2.0 SP1: Success, with errors. Before running the installer, the verb hung while setting the Windows version to `win2k`, at line **1024** of the log. To get around this, I started the task manager with **`export WINEARCH=win32 && export WINEPREFIX=~/.wine.2 && wine taskmgr`**, and closed every process manually, with the task manager itself last. **Like the .NET 2.0 SP2 installation, after ending everything but the task manager, the task manager crashed.**
- XPS Essentials Pack: Success.
- .NET 3.0 SP1: Success.
- .NET 3.5: Success.
- .NET 2.0 SP2: Failure. While running the installer, at line **2024** of the log, the Microsoft .NET Services Installation Utility encountered a fatal error. I clicked the `Don't Send` button. After this, the installation started removing files, and when finished opened up another fatal error message box, at line 2019 of the log. I clicked the `Don't Send Report` button. The main installation window now showed the error text, and link to the error log, saying that the .NET 2.0 runtime was not detected.

## Third Attempt
[Wineboot Log](https://github.com/Winetricks/winetricks/files/1502560/wineboot.log)
[Winetricks Log](https://github.com/Winetricks/winetricks/files/1502561/winetricks-dotnet35sp1.log)
### Program Versions
- `wine`: **`wine-2.21 (Staging)`**
- `winetricks`: `20171018`

### Installations
- .NET 2.0: Success.
- .NET 3.0: Success.
- .NET 2.0 SP1: Success, with errors. Before running the installer, the verb hung while setting the Windows version to `win2k`, at line **799** of the log. **Additionally, `mscorsvw.exe` encountered a fatal error. After clicking `Close`, the verb continued without me having to use the task manager to end the processes.**
![5](https://user-images.githubusercontent.com/13039555/33224286-dd484126-d135-11e7-9ed6-e65dbc0ac2e7.png)
[mscoreswv.exe Backtrace](https://github.com/Winetricks/winetricks/files/1502562/mscoreswv.exe.Backtrace.txt)
- XPS Essentials Pack: Success.
- .NET 3.0 SP1: Success.
- .NET 3.5: Success.
- .NET 2.0 SP2: **Success, with errors. Around the time the installer started, at approximately line 1637 of the log, an unidentified program encountered a fatal error.**
[Unidentified Backtrace](https://github.com/Winetricks/winetricks/files/1502563/Unidentified.Backtrace.txt)
- .NET 3.5 SP1: **Success, with errors. During the installation, an error was logged to the terminal (And not the log file.): `*** Error in 'C:\windows\Microsoft.NET\Framework\v3.5\WFServicesReg.exe': corrupted size vs. prev_size: 0x7c803a8b ***`. (Note: backtick edited to single quote.)**

## Fourth Attempt
[Winetricks Log](https://github.com/Winetricks/winetricks/files/1502564/winetricks-dotnet35sp1.log)
### Program Versions
- `wine`: `wine-2.21 (Staging)`
- `winetricks`: `20171018`

### Installations
- .NET 2.0: Success.
- .NET 3.0: **Failure. After opening the command prompts to silently install the runtime, at line 1595, the verb hung with no immediately apparant reason/error.**

## Fifth Attempt
[Winetricks Log](https://github.com/Winetricks/winetricks/files/1502565/winetricks-dotnet35sp1.log)

Because attempt 4 was such an outlier, the differences of this attempt will be based on attempt 3 instead.
### Program Versions
- `wine`: **`wine-2.21 (Staging)`**
- `winetricks`: `20171018`

### Installations
- .NET 2.0: Success.
- .NET 3.0: **Success, with errors. Somewhere during this verb, an unidentified program encountered a fatal error.**
[Unidentified Backtrace](https://github.com/Winetricks/winetricks/files/1502566/Unidentified.Backtrace.txt)
- .NET 2.0 SP1: **Success.**
- XPS Essentials Pack: Success.
- .NET 3.0 SP1: Success.
- .NET 3.5: Success.
- .NET 2.0 SP2: Success. **Unlike attempt 3, no errors were logged to the terminal.**

**Comment by nabildanial** [comment]
I have this issue and its driving me up the wall. I have a program that uses .NET 3.5 SP1 but I refuse to use Windows.

**Comment by TheKoopaKingdom** [comment]
I haven't heard anything more about this issue, but it might be worth it for you to try using Mono. It doesn't always work for everything that Wine can run after running the .NET installers, but there are cases of either working.

**Comment by qwertychouskie** [comment]
One thing I noticed is that the freeze on setting the win version can be gotten past by killing all of the `mscoreesvc.exe` programs in Task manager/System monitor.

**Comment by qwertychouskie** [comment]
Also, I should mention that I just successfully installed .Net 3.5 SP1 (and 4.6.2) in a fresh Win32 prefix on Wine Staging 2.11 this morning.

**Comment by TheKoopaKingdom** [comment]
Gave Staging 2.11 a few times.
- First attempt: [It](https://i.imgur.com/j0yEGmx.png) [crashed](https://i.imgur.com/0Kannob.png). I wasn't forwarding the log output for this so I just moved on. For this, I forgot to clear the winetricks download cache.
- Second attempt: It hung with a wait timing out. This time, I *did* remove the download cache beforehand, and I did get the full log. [Log link](https://drive.google.com/file/d/0Bz0nv9lp4-39eWJNcVJCMUR6THc/). This log didn't get truncated or anything, it just hung at that `err:ntdll:RtlpWaitForCriticalSection section 0x7bd10500 "../../../wine-staging/dlls/ntdll/virtual.c: csVirtual" wait timed out in thread 0156, blocked by 015c, retrying (60 sec)` line.

Will edit with my further attempts.

**Comment by slyscorpion** [comment]
@TheKoopaKingdom you might want to file a bug in Bugzilla https://bugs.winehq.org/ about that unimplemented function for mscoree.dll. If it gets implemented it might help to alleviate some of the issues you are having with dotnet35sp1...

**Comment by austin987** [comment]
Please don't file a bug for the missing mscoree function, that's not the issue. It simply means that you're trying to use builtin mscoree with native .Net.

The hang is a bug, but I'm not sure why it's hanging. It's also not affecting everyone, which is curious.

**Comment by TheKoopaKingdom** [comment]
I got it to work :D.

So, here's a recap of what I've concluded so far (But take it with a grain of salt, though, because there might be some other conditions that have messed with my results.):
 - If the log is being forwarded, the .NET 3.0 installer will hang at `err:ntdll:RtlpWaitForCriticalSection section 0x7bd10500 "../../../wine-staging/dlls/ntdll/virtual.c: csVirtual" wait timed out in thread 0156, blocked by 015c, retrying (60 sec)`. [Winetricks log](https://drive.google.com/file/d/0Bz0nv9lp4-39eWJNcVJCMUR6THc/).
 - If the log is not being forwarded, when setting the Windows version to `win2k`, a Wine crash occurs ([Crash details](https://pastebin.com/3kMWfnPD), [Winetricks log up to first crash](https://pastebin.com/tkYsWF3C).). Another crash occurs somewhere, but Winetricks just moves right on anyways. One more crash occurs for WFServicesReg.exe, [crash details](https://pastebin.com/JRXM2dm5).. Despite all of these crashes, though, `dotnet35sp1` actually does install.

**Comment by TheKoopaKingdom** [comment]
So, I tested this without log forwarding yesterday, and it still hung, so it seems that the log forwarding is just a coincidence. However, getting around the hang is actually possible.

>One thing I noticed is that the freeze on setting the win version can be gotten past by killing all of the `mscoreesvc.exe` programs in Task manager/System monitor.

It turns out this can also be used to get around the hang. What I did is, in a new terminal tab, I set the Wine prefix to the once that Winetricks was currently installing `dotnet35sp1`, and ran `wine taskmgr`. From there, I ended all three of the `mscoreesvc.exe` processes. After this, the installation continued.

**Comment by sl1pkn07** [comment]
related https://github.com/Winetricks/winetricks/issues/720#issuecomment-274355948

dotnet only install if the wine prefix is a pure 32bits prefix

**Comment by TheKoopaKingdom** [comment]
The installation is broken again. During the installation of the .NET 2.0 SP2 runtime, a fatal error messagebox comes up. Behind it the installation progress bar continues, but when it reaches the end nothing happens. When I close the error, the main window recognizes the error and starts removing the installation files. Here's the error log:
```
[11/01/17,21:30:26] Microsoft .NET Framework 2.0a: [2] Error: Installation failed for component Microsoft .NET Framework 2.0a. MSI returned error code 1603
[11/01/17,21:31:26] WapUI: [2] DepCheck indicates Microsoft .NET Framework 2.0a is not installed.
```

**Comment by austin987** [comment]
This seems to be a wine-staging specific issue. Without staging, I don't get any crash dialogs (though it does now hang for me with a windbg process that never completes. With staging, I see the mscoree crashes you see, as well as a dobule-free in libxml2.

@slackner do you have any thoughts?

**Comment by TheKoopaKingdom** [comment]
I actually gathered [tons of test data](https://i.imgur.com/waeSiyK.png) to figure something out, but it turns out I was on an outdated version of Wine, 2.18. I still have to test on the latest stable version, 2.20.

**Comment by TheKoopaKingdom** [comment]
Added lots of test data to the OP. I'm going to be reinstalling this OS soon, so if anymore data is needed, there may be some variation.

-------
Since this can be quite finicky at times, if anyone else feel like testing it for themselves, here's a template:
```md
## Third Attempt
### Program Versions
- `wine`: 
- `winetricks`: 

### Installations
- .NET 2.0: 
- .NET 3.0: 
- .NET 2.0 SP1: 
- XPS Essentials Pack: 
- .NET 3.0 SP1: 
- .NET 3.5: 
- .NET 2.0 SP2: 
- .NET 3.5 SP1: 
```

**Comment by TheKoopaKingdom** [comment]
I won't be making a full testcase because it's getting pretty redundant, but I did a quick test with Wine 2.22 stable. The SP2 installer still failed.

**Comment by austin987** [comment]
@TheKoopaKingdom at this point I'd recommend filing a wine bug and cc'ing me

**Comment by TheKoopaKingdom** [comment]
Alright, will do when I get a chance.

**Comment by mem-** [comment]
This is probably well known, I got it working with an older version of the Winetricks script.

I'm running Wine 3.6.0 Staging on Debian Jessie (package version 3.6.0~jessie)
Together with Winetricks 0.0+20180217-1 (debian Sid package as of April 2018 pointed out in the Installation section at this site) I get crashes and problems described earlier.

I just wiped my WINEPREFIX (rm -rf $WINEPREFIX) and started from scratch. I had an old Winetricks (version 20141130) on another computer that I copied over to this Jessie machine.

Now .NET 3.5 SP1 installs as expected together with Wine 3.6.0 Staging. So some problems as been introduced since version 20141130. So by trying versions after that you would find in what version of Winetricks the problems started.

**Comment by patdavid** [comment]
Has there been any further movement on this by chance?  I'm using Ubuntu 18.04 with the distro packaged wine 3.0, winetricks 20180217.  I'm hanging up around here doing a `WINEARCH=win32 WINEPREFIX=~/.wineADE winetricks -q dotnet35sp1`:

```
006d:err:ole:create_server class {a9e69610-b80d-11d0-b9b9-00a0c922e750} not registered
006d:err:ole:CoGetClassObject no class object {a9e69610-b80d-11d0-b9b9-00a0c922e750} could be created for context 0x15
006d:err:eventlog:ReportEventW L"WMI classes are not installed."
------------------------------------------------------
Running /usr/bin/wineserver -w. This will hang until all wine processes in prefix=/home/pat/.wineADE terminate
------------------------------------------------------
Executing w_do_call dotnet20sp1
Executing load_dotnet20sp1 
Executing w_do_call remove_mono
Executing load_remove_mono 
------------------------------------------------------
Mono does not appear to be installed.
------------------------------------------------------
------------------------------------------------------
Working around wine bug 34803 
------------------------------------------------------
reg: The system was unable to find the specified registry key or value
reg: The system was unable to find the specified registry key or value
reg: The system was unable to find the specified registry key or value
Executing rm -f /home/pat/.wineADE/dosdevices/c:/windows/system32/mscoree.dll
Executing w_do_call dotnet20
dotnet20 already installed, skipping
------------------------------------------------------
Setting windows version so installer works
------------------------------------------------------
------------------------------------------------------
Running /usr/bin/wineserver -k. This will kill all running wine processes in prefix=/home/pat/.wineADE
------------------------------------------------------
Setting Windows version to win2k
Executing wine regedit /S C:\windows\Temp\_dotnet20sp1\set-winver.reg
------------------------------------------------------
Running /usr/bin/wineserver -w. This will hang until all wine processes in prefix=/home/pat/.wineADE terminate
------------------------------------------------------
```

I'm not sure what a good path forward might be.  @mem- mentions success with 20141130 winetricks, but not sure where to pick that up.

**Comment by CodingKoopa** [comment]
Yep, that's an issue, documented in the first test case at the top:
> To get around this, I started the task manager with `export WINEARCH=win32 && export WINEPREFIX=~/.wine.1 && wine taskmgr`, and ended every process manually, with the task manager itself last.

**Comment by patdavid** [comment]
Yes, I did try that same approach but killing the processes didn't get me past the stall on my side.

I've had some success with rolling back to a previous winetricks as mentioned, but not sure if it all worked ok or if there's further issues with the program I am targeting through all of this.  Thanks for all of the input so far!

**Comment by austin987** [comment]
In a quick test, running `ps aux | grep winedbg` will show the hanging process, which can then be killed via `kill`.

That didn't succeed for me every time. It seems there are two bugs, the crash dialog not showing, and the crash itself. I filed an upstream bug for the first issue, https://bugs.winehq.org/show_bug.cgi?id=45149

**Comment by patdavid** [comment]
I ended up being able to install without issues by using the full package of dotnet 3.5 vs the smaller installer. 

I had to disconnect from the internet to avoid having it hang at possibly Lang pack downloads? Not 100% sure. 

Using the full package and being offline, it installed without any problems.

**Comment by bjmc** [comment]
I'm seeing the same issue as @patdavid, what did you do to get the standalone installer to run successfully? Did you pre-install any winetricks verbs?

**Comment by patdavid** [comment]
@bjmc - I wrote up a small walkthrough of what I did: https://patdavid.net/2018/05/installing-adobe-digital-editions-on-linux-with-wine/#installing

It was for getting Adobe Digital Editions running, but all of the first steps are to get to a good dotnet3.5SP1 install...

**Comment by bjmc** [comment]
Funnily enough, I'm also trying to get ADE working.

**Comment by zfigura** [comment]
I think this happens because we remove mscoree.dll before installing .NET, but dotnet35 installs two dotnet prerequisites, so the second one will end up removing native mscoree.dll.

Why are we removing mscoree.dll?

**Comment by frafra** [comment]
@patdavid it doesn't seem to work anymore (I got a setup error and I tried multiple times following your guide and your comments on winehq).

**Comment by patdavid** [comment]
@frafra ugh, really?  I can try to take a look at it again fresh when I get some time to amend/update.  Did something change I wonder?

**Comment by frafra** [comment]
@patdavid Yes, I tried hard, but it fails at the very end of the installation. It could be a regression with the newer version if Wine (currently using 3.11 as it is the default for Fedora 28).

**Comment by ikeboy** [comment]
Workaround I found: `winetricks dotnet35sp1` was failing at the point of installing dotnet20sp2. Once it exited, you couldn't run it again because it thinks dotnet20 and dotnet35 conflict or something. However, `winetricks dotnet20sp2` works fine to install it later. So what I did was run `winetricks dotnet35sp1`. and as soon as it finished installing `dotnet20sp1`, I opened a new terminal and ran `winetricks dotnet20sp2`. That installed fine, and then when the first process gets around to it, it finds it already installed and goes through the rest of the process without further issues.

This worked on Ubuntu 18.04, Winetricks 20180217, wine-3.0.2

**Comment by blkeller** [comment]
Installations of dotnet35sp1 via winetricks started working for me again once wine 3.13 come out.

**Comment by zfigura** [comment]
Is this still present in current winetricks (specifically after 431bb18719c4acd134983a198fae6c7e98c3b281, but possibly even earlier as per the last comment)?

**Comment by CodingKoopa** [comment]
I've lost track of it, really, it feels like it has been broken and fixed many times. On `Wine 3.20 Staging`, with Winetricks version `20180815`, the `dotnet35sp1 verb works, without any user intervention. Considering this issue solved, closing.

**Comment by bmaupin** [comment]
Using Ubuntu 18.04 with winetricks provided by apt and none of the workarounds mentioned here worked for me. I manually updated winetricks to the latest version and I was successfully able to install dotnet35sp1.

```
$ /usr/bin/winetricks --version
20180217
$ sudo wget https://raw.githubusercontent.com/Winetricks/winetricks/master/src/winetricks -O /usr/local/bin/winetricks
$ sudo chmod +x /usr/local/bin/winetricks 
$ winetricks --version
20181203-next
```

**Comment by lucifertdark** [comment]
> Using Ubuntu 18.04 with winetricks provided by apt and none of the workarounds mentioned here worked for me. I manually updated winetricks to the latest version and I was successfully able to install dotnet35sp1.
> 
> ```
> $ /usr/bin/winetricks --version
> 20180217
> $ sudo wget https://raw.githubusercontent.com/Winetricks/winetricks/master/src/winetricks -O /usr/local/bin/winetricks
> $ sudo chmod +x /usr/local/bin/winetricks 
> $ winetricks --version
> 20181203-next
> ```

Unless I've completely lost the plot it looks to me like you will now have two different versions of Winetricks installed at the same time, one in /usr/bin & the other in /usr/local/bin, recipe for disaster right there.

**Comment by bmaupin** [comment]
> Unless I've completely lost the plot it looks to me like you will now have two different versions of Winetricks installed at the same time, one in /usr/bin & the other in /usr/local/bin, recipe for disaster right there.

Yeah, I was being lazy because I assumed that /usr/bin/winetricks was provided by the wine package and I'd have to uninstall the whole thing to get rid of it. But I guess it's its own package, so I went ahead and uninstalled it.

**Comment by CodingKoopa** [comment]
This issue appears to have started occurring again, with (but not necessarily only starting with) Wine 4.12.1 and Winetricks version `20190615`:
![image](https://user-images.githubusercontent.com/13039555/61594937-fc819f00-abbe-11e9-984a-bb90ddfc7a6a.png)
Commands ran during each test were:
- `export WINEARCH=win32 WINEPREFIX=~/.local/share/wineprefixes/test`
- `rm -rf ~/.cache/winetricks/ "$WINEPREFIX"`
- `wineboot`
- `winetricks dotnet35sp1`

### Wine 4.12.1
- [Wineboot log](https://pastebin.com/wSB4eVK5)
- [Winetricks log](https://pastebin.com/w7TEN0VG)

### Wine 4.12.1 Staging
- [Wineboot log](https://pastebin.com/NYnnPLaw)
- [Winetricks log](https://pastebin.com/ZqpVM4HB)

**Comment by austin987** [comment]
I'm not able to reproduce; works fine here with wine-4.12.1 / WINEARCH=win32 in both quiet and non-quiet modes. Unfortunately nothing is immediately obvious to me from those logs what's wrong.

**Comment by CodingKoopa** [comment]
To further investigate, I grabbed the [error log](https://pastebin.com/UpggGYqQ) from the dialog (from a 4.12.1 non-staging test, but for this log it seems to be pretty irrelevant). It's an exact match to the log [here](https://github.com/Winetricks/winetricks/issues/810#issuecomment-341293426), leading me to believe that the issue is with the .NET 2.0 SP2 runtime, essentially.

Tests were ran with same commands as before (except for the Winetricks verb), all with Wine 4.12.1 non-staging.

### `dotnet20`
Installing the .NET 2.0 runtime succeeded.
- [Winetricks log](https://pastebin.com/YtgEnGq1)

### `dotnet20sp1`
Installing the .NET 2.0 SP1 runtime failed, with the same error message as before.
- [Winetricks log](https://pastebin.com/DCU3tf6W)

### `dotnet20sp2`
Installing the .NET 2.0 SP2 runtime failed, with the same error message as before.
- [Winetricks log](https://pastebin.com/7vpcyfEL)

With this testing, it seems that the problem starts with the .NET 2.0 SP1 runtime / `dotnet20sp1`. Unfortunately, I'm unsure of what resolved this issue last time.

**Comment by MJGHD** [comment]
Hi, I'm also having the issue with dotnet35sp1 and seems to be failing at the same stage

Wine version - wine-4.12.1
Distro - Void

**dotnet35sp1 logs:**
https://pastebin.com/2ukWbZ8g

**dotnet20sp1 logs:**
https://pastebin.com/v4sh01Tv

**Comment by CodingKoopa** [comment]
For the sake of testing things in different environments, I've gathered some testcases on Debian Buster 10.0 running in QEMU, as well as edited past comments and Pastebin pastes to reflect them having been performed in Arch. I've went back to the testcase format + bolding, as there are a lot of different variables here. For Debian, Wine builds were installed from [here](https://wiki.winehq.org/Debian) as this source has Staging builds in addition to Stable and Development.

## Testcase 1
- Distro: `Debian GNU/Linux 10 (buster)`
- Wine Version: `4.12.1 Staging`
- Winetricks Version: `20181203`
- Wine Mono Installed: No
- Wine Gecko Installed: No
- Winetricks verb: `dotnet35sp1`
- Status: Succeeded
- Winetricks log: https://pastebin.com/fTdpXZ6p

## Testcase 2
- Distro: `Debian GNU/Linux 10 (buster)`
- Wine Version: `4.12.1 Staging`
- Winetricks Version: `20181203`
- Wine Mono Installed: **Yes (Through built in installer)**
- Wine Gecko Installed: No
- Winetricks verb: `dotnet35sp1`
- Status: **Failed**
- Winetricks log: https://pastebin.com/6ZMniahu
- Error log: **https://pastebin.com/rsmAfz8m**

---

## Testcase 1
- Distro: `Arch Linux`
- Wine Version: `4.12.1 Staging`
- Winetricks Version: `20190615`
- Wine Mono Installed: No
- Wine Gecko Installed: No
- Winetricks verb: `dotnet35sp1`
- Status: Failed
- Winetricks log: https://pastebin.com/xRa0wkA6
- Error log: https://pastebin.com/UpggGYqQ

## Testcase 2
- Distro: `Arch Linux`
- Wine Version: `4.12.1 Staging`
- Winetricks Version: `20190615`
- Wine Mono Installed: **Yes (Through built in installer)**
- Wine Gecko Installed: No
- Winetricks verb: `dotnet35sp1`
- Status: Failed
- Winetricks log: https://pastebin.com/NYeK4XtD
- Error log: https://pastebin.com/UpggGYqQ

---

From these results, I initially thought that uninstalling Wine Mono was the key, but this only solves the issue on Debian (as, when Wine Mono *is* present on Debian, the error it causes is a different error than the one we have experienced on Arch and Void). Currently, this leads me to believe it's either something with my particular filesystem arrangement, or something about the rolling-update nature of the distros for which this is broken on.

---
Some other (probably unimportant) random oddities:
- On Arch, when Wine Gecko was installed as a system package, `wineboot ` would ask about Wine Mono. When Gecko was uninstalled, `wineboot` asked about that, and never prompted about Mono. This is the reason why for Arch tests, Mono is installed as a system package. Finally, when Mono is installed as a system package (but not Gecko), Gecko is not prompted for. ¯\\_(ツ)_/¯
- On Arch, sometimes, during `wineboot`, this would occur:
```
--- snip ---
wineserver: could not save registry branch to system.reg : No such file or directory
wineserver: could not save registry branch to user.reg : No such file or directory
Could not load wine-gecko. HTML rendering will be disabled.
wine: configuration in '/home/kyle/.local/share/wineprefixes/test' has been updated.
```
Tests where this occurred were omitted from any of the testcases.

**Comment by Gibbz** [comment]
I cannot get dotnet 3.5 or 3.5sp1 installed on the latest version either!
It used to work fine for me, but now it keeps failing.

Im running Manjaro (arch based)

**Comment by tomchiverton** [comment]
Similar on Ubuntu 19.04 too. Even in a fresh prefix, with WINEARCH=win32, always hangs at 

```
Setting Windows version to winxp
Executing wine regedit C:\windows\Temp\_dotnet20\set-winver.reg
------------------------------------------------------
Running /usr/bin/wineserver -w. This will hang until all wine processes in prefix=/home/falken/.wine-ade-2 terminate
----------------
```
Running taskmgr in another window (with same prefix and arch) and killing everything let's it proceed.

If you keep trying again and again with winetricks dotnet35sp1 (with or without --force as needed) it does seem to keep changing the error message, but never works.

Which makes no sense, and isn't much help, sorry.

One other odd thing is ` wine dotnetfx35.exe ` always downloads things. Even though it's the full 200-and-something meg download version.

**Comment by tomchiverton** [comment]
I was only doing this to install Adobe Digital Editions.
The fix is to remove all the Ubuntu wine packages and just use the ones from winehq ppa. Then install latest winetricks which has an option to install ADE (v4) directly.

Now works fine, opens and downloads books purchased from Kobo. 

Fresh wine prefix, wine32 arch, natch.

So wine 3 is just broken, probably will never be fixed. Stop using it. Wine 4 works. 
And I didn't break whatever magic Steam Proton does while doing it..

**Comment by CodingKoopa** [comment]
> So wine 3 is just broken, probably will never be fixed. Stop using it. Wine 4 works.

To be clear, the other recent reports have all been taken with Wine versions from the 4.x series.

> And I didn't break whatever magic Steam Proton does while doing it..

Steam Proton uses its own Wine prefixes, as well as its own Wine binaries, so there's little chance of the two affecting each other (generally, Steam even bundles its own shared libraries that it depends on too).

**Comment by Whisprin** [comment]
Using Arch Linux I tried several versions of the wine package. Installing dotnet35sp1 failed for versions 4.10 - 4.15, which is at the moment the most recent version.

However, using Wine 4.9 installing dotnet35sp1 did work without a problem. Upgrading wine afterwards still keeps it running. So my workaround for now is downgrade Wine to 4.9, install dotnet35sp1 and upgrade wine again.

I'm using winetricks 20190615 across all Wine versions.

**Comment by 4llan** [comment]
I'm pretty confident that this issue was introduced by a change in wine-msi since 4.11.
https://bugs.winehq.org/show_bug.cgi?id=47724

**Comment by CodingKoopa** [comment]
With that bug being closed, I have performed some additional testing. For all of these, neither Wine Mono nor Wine Gecko were installed.

## Testcase 1
- Distro: `Debian GNU/Linux 10 (buster)`
- Wine Version: `wine-4.0.3` (`winehq-stable`)
- Winetricks Version: `20181203`
- Winetricks verb: `dotnet35`
- Status: Success
- Winetricks log: https://pastebin.com/qmGyPDu2

## Testcase 2
- Distro: `Debian GNU/Linux 10 (buster)`
- Wine Version: **`wine-5.0-rc2` (`winehq-devel`)**
- Winetricks Version: `20181203`
- Winetricks verb: `dotnet35`
- Status: Success
- Winetricks log: **https://pastebin.com/Yn1Afqce**

## Testcase 3
- Distro: `Debian GNU/Linux 10 (buster)`
- Wine Version: **`wine-5.0-rc2 (Staging)` (`winehq-staging`)**
- Winetricks Version: `20181203`
- Winetricks verb: `dotnet35`
- Status: Success
- Winetricks log: **https://pastebin.com/7SNiLjtk**

---

## Testcase 1
- Distro: `Arch Linux`
- Wine Version: `wine-4.0.3` ([`wine-stable`](https://aur.archlinux.org/packages/wine-stable/))
- Winetricks Version: `20191224`
- Winetricks verb: `dotnet35`
- Status: Failure
![image](https://user-images.githubusercontent.com/13039555/71502464-4ccfed80-283e-11ea-9182-3d5d78385fcc.png)
- Winetricks log: https://pastebin.com/zigjQaem
- Error Log 1: https://pastebin.com/J96xwjgR
- Error Log 2: https://pastebin.com/ke1A5c9j

## Testcase 2
- Distro: `Arch Linux`
- Wine Version: **`wine-5.0-rc2` ([`wine`](https://www.archlinux.org/packages/multilib/x86_64/wine/))**
- Winetricks Version: `20190615`
- Winetricks verb: `dotnet35sp1`
- Status: Failure
- Winetricks log: **https://pastebin.com/5UWituK6**
- Error Log 1: https://pastebin.com/x9Uuxhqg
- Error Log 2: https://pastebin.com/qqHcwdcq

## Testcase 2
- Distro: `Arch Linux`
- Wine Version: **`wine-5.0-rc2 (Staging)` ([`wine-staging`](https://www.archlinux.org/packages/multilib/x86_64/wine-staging/))**
- Winetricks Version: `20190615`
- Winetricks verb: `dotnet35sp1`
- Status: Failure
- Winetricks log: **https://pastebin.com/xPNVWSda**
- Error Log 1: https://pastebin.com/6twt8Fc0
- Error Log 2: https://pastebin.com/7TC2qVEf

---

This data shows that installation of `dotnet35` (which should more or less produce the same results as `dotnet35sp1`) is now successful on Debian, but not on my Arch Linux installation. I'm unsure of what is causing this difference. Cheers, hope you're all doing well 😃

**Comment by austin987** [comment]
A) does installing winbind on arch help (I realize debian is also missing, but IIRC it has broken things for some people before).
B) https://blogs.msdn.microsoft.com/aaronru/2008/05/21/when-installing-net-framework-3-5-on-vista-what-does-the-net-framework-2-0-sp1-on-vista-error-code-1058-mean-and-how-do-i-work-around-it/ implies that your arch error is due to the windows update service being disabled. Wine has a basic implementation of that, though I don't know why Arch fails there.

My suggestion would be to compare the differences in the two systems. I.e., if you build wine yourself on both, does it work? Are they using different compilers? Does arch apply custom patches?

**Comment by CodingKoopa** [comment]
> Does arch apply custom patches?

Going by the `prepare()` function in the Arch [`wine` PKGBUILD script](https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/wine), not really. I'm less familiar with the Debian build system, but I've installed my package from [WineHQ](https://wiki.winehq.org/Debian), so I doubt that patches have been applied here. I wasn't able to conclusively determine this myself, because the OBS confuses me.

I will try the other things you've listed when I get the chance, cheers.

**Comment by mirh** [comment]
Put aside that I had to kill `mscorsvw` half way through the install procedure, it seems to install just fine for me on manjaro.

**Comment by CodingKoopa** [comment]
What Wine version (`wine --version`) are you using?

**Comment by mirh** [comment]
5.5 staging

**Comment by CodingKoopa** [comment]
`dotnet35` and `dotnet35sp1` both work for me now. Tested on Arch, with `wine` and `wine-stable`. I also have reports of it working with `wine-staging`.

**Comment by Pockets-byte** [comment]
having the same issue, fedora 40, wine 9.5
winetricks fails in the described way when trying to install dotnet35 and dotnet35sp1

**Comment by Chiitoo** [comment]
Pockets-byte,

> having the same issue, fedora 40, wine 9.5 winetricks fails in the described way when trying to install dotnet35 and dotnet35sp1 this is on a fresh 32 bit prefix

Please verify that you're using a reasonably up-to-date version of Winetricks (current release is '20240105'), and if the issue persist, would be good to try the current 'git master' version as well, which you can simply copy from the repository.

If the problem still persists, please open a new issue for it (this one was closed in 2020 as being solved).

As a side note, 9.5 is a development release, of which the currently latest one is at 9.12.

The current stable release is 9.0.
